test/advection_test/test_host.F90

```diff
@@ -314,19 +318,63 @@ subroutine test_host(retval, test_suites)
       if (.not. is_constituent) then
          write(6, *) "ERROR: cloud_ice_dry_mixing ratio not found in ",       &
                         "host cap constituent list"
-         errflg_final = -1 !Notify test script that a failure occurred
+         errflg_final = -1 ! Notify test script that a failure occurred
       end if
 
 
       ! Register the constituents to find out what needs advecting
+      ! DO A COUPLE OF TESTS FIRST
+
+      ! First confirm the correct error occurs if you try to add an
+      ! incompatible constituent with the same standard name
+      expected_error = 'ccp_model_const_add_metadata ERROR: Trying to add ' //&
+              'constituent specific_humidity but an incompatible ' //         &
+              'constituent with this name already exists'
+      allocate(host_constituents(2))
       call host_constituents(1)%instantiate(std_name="specific_humidity",     &
            long_name="Specific humidity", units="kg kg-1",                    &
            vertical_dim="vertical_layer_dimension", advected=.true.,          &
+           min_value=1000._kind_phys, molar_mass=2000._kind_phys,             &
+           errcode=errflg, errmsg=errmsg)
+      call host_constituents(2)%instantiate(std_name="specific_humidity",     &
+           long_name="Specific humidity", units="kg kg",                      &
+           vertical_dim="vertical_layer_dimension", advected=.true.,          &
+           min_value=1000._kind_phys, molar_mass=2000._kind_phys,             &
+           errcode=errflg, errmsg=errmsg)
+      call check_errflg(subname//'.initialize', errflg, errmsg, errflg_final)
+      if (errflg == 0) then
+         call test_host_ccpp_register_constituents(host_constituents,         &
+                 errmsg=errmsg, errflg=errflg)
+      end if
+      ! Check the error 
+      if (errflg == 0) then
+         write(6, '(2a)') 'ERROR register_constituents: expected this error: ', &
+                 trim(expected_error)
+      else
+         if (trim(errmsg) /= trim(expected_error)) then
+            write(6, '(4a)') 'ERROR register_constituents: expected this error: ', &
+                 trim(expected_error), ' Got: ', trim(errmsg)
+         end if
+      end if
+      ! Now try again but with a compatible constituent - should be ignored when
+      ! the constituents object is created
+      call test_host_ccpp_deallocate_dynamic_constituents()
+      deallocate(host_constituents)
+      allocate(host_constituents(2))
+      call host_constituents(1)%instantiate(std_name="specific_humidity",     &
```

Comment:
I probably don't understand the test here, but how is it allowed to have two host constituents with the same standard_name?

Reply:
It's allowed if they're identical (in which case it's just not added at all). This handles the case where two different schemes dynamically add the same constituent.